src/huggingface_hub/hf_api.py

```diff
@@ -264,6 +269,121 @@ def delete_repo(
         )
         r.raise_for_status()
 
+    def upload_file(
+        self,
+        path_or_fileobj: Union[str, BinaryIO],
+        path_in_repo: str,
+        repo_id: str,
+        repo_type: Optional[str] = None,
+        revision: Optional[str] = None,
+        token: Optional[str] = None,
+    ) -> str:
+        """
+        Upload a local file (up to 5GB) to the given repo, tracking it with LFS if it's larger than 10MB
+
+        Params:
+            filepath (:obj:`os.Pathlike`):
+                Path to the file on the local machine, for instance: :obj:`"~/my_models/weights.bin"`
+
+            path_in_repo (:obj:`str`):
+                Relative filepath in the repo, for example: :obj:`"checkpoints/1fec34a/weights.bin"`
+
+            repo_id (:obj:`str`):
+                The repository to which the file will be uploaded, for example: :obj:`"username/custom_transformers"`
+
+            repo_type (:obj:`str`, Optional):
+                Set to "dataset" if uploading to a dataset, default is model.
+
+            revision (:obj:`str`, Optional):
+                The git revision to commit from. Defaults to the :obj:`"main"` branch.
+
+            token (:obj:`str`, Optional):
+                Authentication token, obtained with :function:`HfApi.login` method
+
+        Returns:
+            :obj:`str` : The URL to visualize the uploaded file on the hub
+
+        Raises:
+            :class:`ValueError` : if some parameter value is invalid
+
+            :class:`requests.HTTPError` : if the HuggingFace API returned an error
+
+        Examples:
+            >>> with open("./local/filepath", "rb") as fobj:
+            ...     upload_file(
+            ...         path_or_fileobj=fileobj,
+            ...         path_in_repo="remote/file/path.h5",
+            ...         repo_id="username/private-dataset",
+            ...         repo_type="datasets",
+            ...         token="my_token",
+            ...    )
+            "https://huggingface.co/datasets/username/private-dataset/blob/main/remote/file/path.h5"
+
+            >>> upload_file(
+            ...     path_or_fileobj=".\\\\local\\\\file\\\\path",
+            ...     path_in_repo="remote/file/path.h5",
+            ...     repo_id="username/private-model",
+            ...     token="my_token",
+            ... )
+            "https://huggingface.co/username/private-model/blob/main/remote/file/path.h5"
```

Comment:
```suggestion
            >>> upload_file(
            ...     path_or_fileobj=".\\\\local\\\\file\\\\path",
            ...     path_in_repo="remote/file/path.h5",
            ...     repo_id="username/my-model",
            ...     token="my_token",
            ... )
            "https://huggingface.co/username/my-model/blob/main/remote/file/path.h5"
```
(nit) so that people don't end up thinking it's only for private repos? 😅